### ISSUE TITLE
[Shapes] Wrong Polygon Contains behaviour

### ISSUE THREAD

**Issue by Spessi** [craftworkgames/MonoGame.Extended#214]
Hi,
due to the missing comments in the PolygonF class, I'm not sure if the Contains method behaves correctly or incorrectly.

Here's a code example:

``` C#
// Create 4 vertices, representing a square, starting at (10/10) and ending at (20/20)
List<Vector2> testVertices = new List<Vector2>();
testVertices.Add(new Vector2(10, 10));
testVertices.Add(new Vector2(10, 20));
testVertices.Add(new Vector2(20, 20));
testVertices.Add(new Vector2(20, 10));

// Create a polygon out of the previously created vertices
PolygonF testPoly = new PolygonF(testVertices);

// Let's do some edge tests
bool res;
res = testPoly.Contains(10, 10); // true
res = testPoly.Contains(20, 20); // false
```

What I'd expected is that both results are true or false (in my case, both true would be better).

I'll now try to understand the algorithm and fix it, so that both test cases will be true. If you confirm it's a bug and not a wanted behaviour, I'd try to submit a PR (it's my first time :-)).

**Comment by craftworkgames** [comment]
To consistent with #213 we want `Contains` to behave the same way it does on standard MonoGame shapes like `Rectangle`. I'm not certain what this behavior actually is so if you want to investigate that'd be great.

Unfortunately, when returning a `bool` for something that actually has 3 states (contains, not contains, touching) it's always going to be ambiguous. That said, MonoGame is the source of truth here, so we need to match it.

> (in my case, both true would be better).

May I ask what you using these for out of curiosity?

**Comment by charles-esterbrook** [comment]
> Unfortunately, when returning a bool for something that actually has 3 states (contains, not contains, touching) it's always going to be ambiguous. That said, MonoGame is the source of truth here, so we need to match it.

Should we have an additional method that returns the relation?

`enum Relation { Outside, Inside, Touches };

**Comment by craftworkgames** [comment]
> Should we have an additional method that returns the relation?

Yeah we could do that. I'm pretty sure it's common practice in physics engines. It might be worth taking a look at how [Farseer](http://farseerphysics.codeplex.com/) does it or one of the others. I don't think they call it a relation.

**Comment by craftworkgames** [comment]
Any one want to take a closer look at this bug?

**Comment by LithiumToast** [comment]
I'll get around to this with collisions. Though, I will be gutting the `PolygonF` struct in favour of a `Polygon` class.

**Comment by craftworkgames** [comment]
@LithiumToast have you already finished the new shape classes? It might be worth pulling those in sooner rather than later.

**Comment by lithiumtoast** [comment]
Troubled code is not standard with the [orange book](https://www.amazon.ca/Real-Time-Collision-Detection-Christer-Ericson/dp/1558607323). Will need to look at all the surrounding code for the context of what is good and not so good.

**Comment by AristurtleDev** [comment]
Adding reference for bookeeping, `PolygonF` was changed to `Polygon` during the following PR

- https://github.com/craftworkgames/MonoGame.Extended/pull/346

**Comment by AristurtleDev** [comment]
Referencing what @dylanwilson80 said above

> To consistent with https://github.com/craftworkgames/MonoGame.Extended/issues/213 we want Contains to behave the same way it does on standard MonoGame shapes like Rectangle. 

I agree that consistency in behavior should exist between MonoGame and MonoGame.Extended in this regard.  In MonoGame, you can observe if a point is "contained" at an edge with the following:

```cs
Rectangle rect = new Rectangle(0, 0, 20, 20);
var contains = rect.Contains(new Point(20, 20));
Console.WriteLine(contains);

// outputs false
```

Which will output `false`.  This means the original issue is a non-issue and this can be closed.  Will add a unit test to ensure consistency in behavior before closing.

**Comment by AristurtleDev** [comment]
Closing this as resolved with PR #871 
@Spessi if you would like to continue discussion on a feature request for returning state of contains as an enum as discussed above, please open a new issue.
